sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections;
+
+namespace System.ClientModel.Primitives;
+
+internal abstract class CollectionWriter
+{
+    internal static CollectionWriter GetCollectionWriter(IEnumerable enumerable, ModelReaderWriterOptions options)
+    {
+        if (options.Format != "J" && options.Format != "W")
+        {
+            throw new InvalidOperationException($"Format '{options.Format}' is not supported only 'J' or 'W' format can be written as collections");
+        }
+
+        if (options.Format == "J")
+        {
+            return new JsonCollectionWriter();
+        }
+        else // W format
+        {
+            var persistableModel = GetIPersistableModel(enumerable);
+            var wireFormat = persistableModel.GetFormatFromOptions(options);
+            if (wireFormat == "J" && persistableModel is IJsonModel<object>)
+            {
+                return new JsonCollectionWriter();
+            }
+            throw new InvalidOperationException($"{persistableModel.GetType().Name} has a wire format of '{wireFormat}' it must be 'J' to be written as a collection");
+        }
+    }
+
+    private static IPersistableModel<object> GetIPersistableModel(IEnumerable enumerable)
+    {
+        object first = GetFirstObject(enumerable);
+        if (first is IEnumerable nextEnumerable)
+        {
+            return GetIPersistableModel(nextEnumerable);
+        }
+        else if (first is IPersistableModel<object> persistableModel)
+        {
+            return persistableModel;
+        }
+        else
+        {
+            throw new InvalidOperationException($"Unable to write {enumerable.GetType().Name} can only write collections of IPersistableModel");
+        }
+    }
+
+    private static object GetFirstObject(IEnumerable enumerable)
+    {
+        if (enumerable is IDictionary dictionary)
+        {
+            foreach (var key in dictionary.Keys)
+            {
+                var value = dictionary[key];
+                if (value is not null)
+                {
+                    return value;
+                }
+            }
+            throw new InvalidOperationException($"Can't use format 'W' format on an empty collection please specify a concrete format");
```

Comment:
```suggestion
            throw new InvalidOperationException($"Can't use format 'W' format on an empty collection. Please specify a concrete format");
```

Reply:
I'm not in love with this message.  I wouldn't equate "a concrete format" with "go add something to this collection."